<html>
<!-- Author: Junko Tsuji -->
<!-- Copyright 2012, Junko Tsuji -->
<!-- Creation Date: 2012.12.28 -->
<head>
<meta http-equiv="Content-type" content="text/html; charset=utf-8">
<meta keywords="numt, nupt, organelle, nuclear, genome">
<title>norg-suite</title>
<body leftmargin="50" topmargin="50">

<center>
<h1>
<font size="8" color="#D02090" face="Comic Sans MS">
norg-suite
</font>
</h1>
<font size="5" color="#DA70D6"><b>n</b></font>uclear
<font size="5" color="#DA70D6"><b>org</b></font>anelle DNA
<font size="5" color="#DA70D6"><b>suite</b></font>
<hr>
</center>

<h2>Organelle DNA insertions?</h2>
It is known that most eukaryotic nuclear genomes contain regions derived
from organelle DNA insertions, so-called <b>NUMT</b>s and <b>NUPT</b>s
(<b>NU</b>clear <b>M</b>i<b>T</b>ochondrial and <b>P</b>las<b>T</b>idal sequences).
<center>
<img src="./norg-suite-pic.jpg" height="33%" width="35%">
</center>

<h2>What's norg-suite?</h2>
<font color="#D02090" size="4"><b>norg-suite</b></font>
 is a software package which accurately delineates
 organelle DNA insertions in nuclear genomes.<br><br>
<a href="">Download</a>

<h2>Contents</h2>
<font color="#D02090" size="4"><b>norg-suite</b></font>
contains two programs:<br>
<ul>
    <li><h3>
    <a href="./norg-seq.html"><font color="#4682B4"><b>norg-seq</b></font></a>
    </h3></li>
        Finds organelle genome insertions in the nuclear genomes
        with local alignments between the organelle and the nuclear genomes.
    <li><h3>
    <a href="./norg-age.html"><font color="#FFA500"><b>norg-age</b></font></a>
    </h3></li>
        Estimates evolutionary ages of organelle insertions defined by
        <font color="#4682B4"><b>norg-seq</b></font>,
        with phylogenetic trees and genome alignments.
</ul>

<h2>Requirement</h2>
<font color="#D02090" size="4"><b>norg-suite</b></font>
 requires the following environment to run:<br>
<ol>
    <li><a href="http://www.python.org/download/">
        Python</a> 2.5, or more latest</li>
    <br>
    <li>Third party tools</li>
    <br>
    <ul>
    <li><font color="#4682B4"><b>norg-seq</b></font> needs</li>
      <ul>
      <li><a href="http://www.cbrc.jp/tantan/">tantan</a>
          - Repeat masking software</li>
      <li><a href="http://last.cbrc.jp/">LAST</a>
          - Local alignment software</li>
      </ul>
    <br>
    <li><font color="#FFA500"><b>norg-age</b></font> needs</li>
      <ul>
      <li><a href="http://mafft.cbrc.jp/alignment/software/">MAFFT</a>
          - Multiple alignment software</li>
      <li><a href="http://sco.h-its.org/exelixis/software.html">RAXML</a>
          - Phylogenetic tree inference software</li>
      </ul>
</ol>
<ul>
    To prepare the above tools easily,
    <font color="#D02090"><b>norg-suite</b></font> provides a shell script
    which automatically downloads the latest versions of the above tools.
    <br>
    Under "tools" directory, you can use the script like this:
    <ul>
    <table border="0" bgcolor="#E6E6FA" width="95%">
    <br>
    <td>
    <pre>

  $ chmod +x fetch.sh
  $ fetch.sh -a
    </pre>
    </td>
    </table>
    </ul>
    <br>
    After fetching the tools, please follow each software instruction to compile.
</ul>

<h2>Publication</h2>
<ul>
    <li><a href="http://nar.oxfordjournals.org/content/40/18/9073.abstract">
        Mammalian NUMT insertion is non-random.</a>
        J Tsuji, MC Frith, K Tomii, P Horton. Nucleic Acids Research 2012.</li>
    <li>Automatic Accurate Delineation of Organelle DNA Copies in the Nuclear Genomes.
        J Tsuji, MC Frith, P Horton. (<i>in preparation</i>)</li>
</ul>

<h2>Contact</h2>
If you have problems please let us know:
<ul>
    <table border="0" cellspacing="4">
        <td>Package developer</td>
        <td><pre>  tsuji (at) k.u-tokyo.ac.jp</pre></td>
    <tr>
        <td>Corresponding author</td>
        <td><pre>  horton-p (at) aist.go.jp</pre></td>
    </table>
</ul>

<hr>
<br>
<center><a href="./norg-suite.html">top</a></center>
</body>
</html>
