<html>
<!-- Author: Junko Tsuji -->
<!-- Copyright 2012, Junko Tsuji -->
<!-- Creation Date: 2012.12.28 -->
<head>
<meta http-equiv="Content-type" content="text/html; charset=utf-8">
<meta keywords="numt, nupt, organelle, nuclear, genome">
<title>norg-age usage</title>
<body leftmargin="50" topmargin="50">

<center>
<h1>
<font size="8" color="#FF6347" face="Comic Sans MS">
norg-age
</font>
</h1>
<font size="5" color="#FFA500"><b>n</b></font>uclear
<font size="5" color="#FFA500"><b>org</b></font>anelle DNA insertion
<font size="5" color="#FFA500"><b>age</b></font> estimator
<hr>
</center>

<h2>Usage</h2>
<font color="#FF6347" size="4"><b>norg-age</b></font> runs with:
<ol>
    <li>Pathes of MAFFT and RAxML executables</li>
    <li>Organelle genome file of target species and closely related species
        (<b>multi fasta</b>)</li>
    <br>
    <font size="4"><b>Note</b></font>
    <ul>
        <li>An organelle genome of target species must be
            placed at the top of all sequences</li>
        <li>All organelle genomes must start a homologous position
            (e.g. starting at the D-loop origin)</li>
    </ul>
    <br>
    <li>Outgroup name of the above organelle genomes</li>
    <li>Organelle DNA insertion sequence file(s) caluclated by
        <font color="#4682B4"><b>norg-seq</b></font></li>
</ol>
You can use it like this:
    <ul><table border="0" bgcolor="#E6E6FA" width="95%">
    <td>
    <pre>

  $ norg-age --mafft PATH --raxml PATH --outgroup NAME organelleGenome fastaFile(s)
    </pre>
    </td>
    </table>
    </ul>

<h2>Output</h2>
<font size="4" color="#FF6347"><b>norg-age</b></font> finally outputs:
<ul>
    <li><h4>Coordinates of organelle DNA insertions with
       <font size="4" color="#FF00FF"><b>age labels</b></font></h4></li>
    <ul>
        This coordinate file is in
        <a href="http://genome.ucsc.edu/FAQ/FAQformat.html#format1">BED format</a>
        (see also: <a href="./norg-seq.html">
                   <font color="steelblue"><b>norg-seq</b></font>'s format</a>
        for more detail),
        and it includes inferred ages labeled with alphabets at forth column
        (i.e. field for organelle DNA coordinates).
        The result file also shows reference tree in
        <a href="http://en.wikipedia.org/wiki/Newick_format">Newick format</a>
        and the detail of labeled ages at the top of the file.<br><br>
        An example of the inferred reference tree and ages looks like this:
        <ul>
        <table border="0" bgcolor="#E6E6FA" width="95%">
        <br>
        <td>
        <pre>

  #
  # Reference tree:
  # (chimpanzee_16554bp,human_16571bp,(gorilla_16364bp,(orangutan_16389bp,(
  #       gibbon_16472bp,(rhesus_16564bp,(squirrel_monkey_16474bp,mouse_16299bp))))));
  #
  # Lables:
  # a: human_16571bp
  # b: gorilla_16364bp,(chimpanzee_16554bp)
  # c: orangutan_16389bp
  # d: gibbon_16472bp
  # e: rhesus_16564bp
  # f: squirrel_monkey_16474bp
  # o: mouse_16299bp
  #
        </pre>
        </td>
        </table>
        </ul>
        <br>
        Each labeled insertion coordinate is color-coded according to each age and
        can be displayed on the <a href="http://genome.ucsc.edu/">UCSC Genome Browser</a>
        as <a href="http://genome.ucsc.edu/goldenPath/customTracks/custTracks.html">
        Custom Tracks</a>.
    </ul>
</ul>

<h2>Options</h2>
<font color="#FF6347" size="4"><b>norg-age</b></font>
can be controlled with following options:

    <h4>Cosmetic Options</h4>
    <table border="0" cellspacing="7">
        <td valign="top"><pre>-h, --help                   </pre></td>
        <td>Show all options and their default settings.</td>
    <tr>
        <td valign="top"><pre>-v, --verbose                </pre></td> 
        <td>Be verbose: print what norg-age is doing.</td>
    </table>

    <h4>Reference Organelle Options</h4>
    <table border="0" cellspacing="7">
        <td valign="top"><pre>--ref-msa FILE</pre></td>
        <td>Precomputed multiple sequence alignments (MSA; in fasta format)
            of reference organelle genomes, to skip the the MSA computaion step.</td>
    <tr>
        <td valign="top"><pre>--ref-tree FILE</pre></td>
        <td>Precomputed phylogenetic tree (in Newick format) of reference organelle
            genomes, to skip the phylogenetic tree calculation step.</td>
    <tr>
        <td valign="top"><pre>--alignment-net FILE  </pre></td>
        <td>List of UCSC "net" (genome pairwise alignment) files with names of
            closely related species in the multi fasta organelle genome file,
            to verify estimated insertion ages (see also:
            <nobr><font face="Courier" size="2">--ignore-card</font></nobr>).
            Each field in the file must be separated by spaces.<br>
            For example, the format targeted human organelle insertions is like this:
            <ul>
            <table border="0" bgcolor="#E6E6FA" width="95%">
            <br>
            <td>
            <pre>

  chimp   /path/to/human.chimp.net
  rhesus  /path/to/human.rhesus.net
  gibbon  /path/to/human.gibbon.net

  ...
            </pre>
            </td>
            </table>
            </ul>
    <tr>
        <td valign="top"><pre>--ignore-card CHR1,CHR2,...  </pre></td>
        <td>Comma-separated chromosome names of target species to exclude the
            specific chromosomes from the alignment net calculation
            (see also: <nobr><font face="Courier" size="2">--alignment-net</font></nobr>)</td>
    <tr>
        <td valign="top"><pre>--genome-seqs FILE</pre></td>
        <td>List of nuclear genome files of all species in the multi fasta
            organelle genome file, to verify estimated insertion ages (see also:
            <nobr><font face="Courier" size="2">--last</font></nobr>).
            Each field in the file must be separated by speces.<br>
            For example, the format is like this:
            <ul>
            <table border="0" bgcolor="#E6E6FA" width="95%">
            <br>
            <td>
            <pre>

  human   /path/to/humanGenome.mfa
  chimp   /path/to/chimpGenome.mfa
  rhesus  /path/to/rhesusGenome.mfa
  gibbon  /path/to/gibbonGenome.mfa

  ...
            </pre>
            </td>
            </table>
            </ul>
    <tr>
        <td valign="top"><pre>--last PATH</pre></td>
        <td>Path to 'last' executables to run the
            '<nobr><font face="Courier" size="2">--genome-seqs</font></nobr>'
            option (see also:
            <nobr><font face="Courier" size="2">--genome-seqs</font></nobr>).</td>
    </table>
<br>
<hr>
<br>
<center>
<a href="./norg-suite.html">back</a>
</center>
